_posts/2023-09-07-develop-functions-locally.md

```diff
@@ -0,0 +1,278 @@
+  let token = await readFile("/var/openfaas/secrets/api-key", )
+
+  const result = {
+    'body': JSON.stringify(event.body),
+    'api-key': token.toString(),
+    'content-type': event.headers["content-type"]
+  }
+
+  return context
+    .status(200)
+    .succeed(result)
+}
+```
+
+Run the function with local-run and see that the secret is mounted in the container and can be read for use within the function.
+
+Run the echo function:
+
+```bash
+faas-cli local-run echo
+```
+
+Invoke the function with curl and inspect the response:
+
+```bash
+$ curl -s http://127.0.0.1:8080 \
+  -H "Content-Type: text/plain" \
+  -d "Greetings from OpenFaaS" | jq
+
+{
+  "body": "\"Greetings from OpenFaaS\"",
+  "api-key": "secret-access-token",
+  "content-type": "text/plain"
+}
+```
+
+You should see that the api-key secret is included in the JSON response.
+
+## Watch for changes
+
+Another way to quickly iterate on function during development is by using the built in watch functionality of the CLI.
+
+The `--watch` flag can be used with both the `local-run` and `up` command. Adding the flag will tell the cli to watch for any changes to the function source files and automatically re-build and deploy functions on save.
+
+## Automatic tagging of function images
+
+All OpenFaaS functions are built into immutable docker images. By default if no image tag is included for a function in the stack.yml file the `:latest` tag is used. When iterating over function and pushing them to an image registry it is a best practice to organise different image versions using tags instead of always pushing to `:latest`.
+
+There are two options to set tags for function images.
+
+1. Set the image tag in the stack.yml file
+    
+    ```
+    image: my-fn:0.1.0
+    ```
+
+    [Environment variable substitution](https://docs.openfaas.com/reference/yaml/#yaml-environment-variable-substitution) could be used to configure the tag with environment variables.
+
+2. Let the faas-cli generate the tag automatically.
+
+    The `--tag` option can be used with the `build`, `push` and `deploy` sub-commands of the faas-cli. If this flag is provided, image tags for functions will automatically be generated based on available metadata. This can be either Git metadata like the commit sha or branch name or a hash of the digest of the function handler content.
+
+    Some examples:
+
+    When using the flag `--tag=sha` the image tag used in the stack.yml file is suffixed with the short Git SHA. e.g
+
+    ```
+    image: my-fn:0.2 => image: my-fn:0.2-cf59cfc
+    ```
+
+    if not tag is set in the stack.yml file the suffix is appended to latest:
+
+    ```
+    image: my-fn => image: my-fn:latest-cf59cfc
+    ```
+
+    Our docs have a detailed overview of [all the available tag versions](https://docs.openfaas.com/cli/tags/).
+
+
+## Conclusion
+To quickly run and test individual OpenFaaS functions locally with minimal overhead, use `faas-cli local-run`.
```

Comment:
We set out to show you how the new `local-run` command worked for faster iteration, the `--watch` flag for live-reloading as you edit, and the `--tag digest` flag to generate dynamic tags as you edit.

When it comes to iterating on functions, doing a full deployment is the easiest and most realistic way to test your work, however, `faas-cli local-run` can speed things up if your function doesn't have a lot of dependencies.